[3.0.1] Test failures in test_sampler.py

**General information:**

- emcee version: 3.0.1
- platform: Debian unstable/testing
- installation method (pip/conda/source/other?): pip source

**Problem description:**

During packaging the new version for Debian, I get the following test errors:
```
_____________________________ test_errors[Backend] _____________________________

backend = <class 'emcee.backends.backend.Backend'>, nwalkers = 32, ndim = 3
nsteps = 5, seed = 1234

    @pytest.mark.parametrize("backend", all_backends)
    def test_errors(backend, nwalkers=32, ndim=3, nsteps=5, seed=1234):
[…]
    
            # Ensure that a warning is logged if the inital coords don't allow
            # the chain to explore all of parameter space, and that one is not
            # if we explicitly disable it, or the initial coords can.
            with pytest.warns(RuntimeWarning) as recorded_warnings:
>               sampler.run_mcmc(np.ones((nwalkers, ndim)), nsteps)

emcee/tests/unit/test_sampler.py:120: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
emcee/ensemble.py:381: in run_mcmc
    for results in self.sample(initial_state, iterations=nsteps, **kwargs):
emcee/ensemble.py:247: in sample
    np.atleast_2d(np.cov(state.coords, rowvar=False))
emcee/ensemble.py:558: in _scaled_cond
    return np.linalg.cond(c.astype(float))
/usr/lib/python3/dist-packages/numpy/linalg/linalg.py:1716: in cond
    s = svd(x, compute_uv=False)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

a = array([[nan, nan, nan],
       [nan, nan, nan],
       [nan, nan, nan]])
full_matrices = True, compute_uv = False

    @array_function_dispatch(_svd_dispatcher)
    def svd(a, full_matrices=True, compute_uv=True):
      """[…]"""
        a, wrap = _makearray(a)
        _assertRankAtLeast2(a)
        t, result_t = _commonType(a)
    
        extobj = get_linalg_error_extobj(_raise_linalgerror_svd_nonconvergence)
    
        m, n = a.shape[-2:]
        if compute_uv
            […]
        else:
            if m < n:
                gufunc = _umath_linalg.svd_m
            else:
                gufunc = _umath_linalg.svd_n
    
    
            signature = 'D->d' if isComplexType(t) else 'd->d'
>           s = gufunc(a, signature=signature, extobj=extobj)
E           ValueError: On entry to DLASCL parameter number 5 had an illegal value
/usr/lib/python3/dist-packages/numpy/linalg/linalg.py:1624: ValueError
```
and the same for `TempHDFBackend`.
Numpy version is 1.16.5, Python 3.7.5, x86_64.
This is a regression; 3.0.0 does not show this failure in the same build environment.

--- a/src/emcee/__init__.py
+++ b/src/emcee/__init__.py
@@ -22,11 +22,12 @@ __description__ = "The Python ensemble sampling toolkit for MCMC"
 from .emcee_version import __version__  # isort:skip
 
 from . import autocorr, backends, moves
-from .ensemble import EnsembleSampler
+from .ensemble import EnsembleSampler, walkers_independent
 from .state import State
 
 __all__ = [
     "EnsembleSampler",
+    "walkers_independent",
     "State",
     "moves",
     "autocorr",

--- a/src/emcee/ensemble.py
+++ b/src/emcee/ensemble.py
@@ -78,13 +78,19 @@ class EnsembleSampler(object):
     ):
         # Warn about deprecated arguments
         if a is not None:
-            deprecation_warning("The 'a' argument is deprecated, use 'moves' instead")
+            deprecation_warning(
+                "The 'a' argument is deprecated, use 'moves' instead"
+            )
         if threads is not None:
             deprecation_warning("The 'threads' argument is deprecated")
         if runtime_sortingfn is not None:
-            deprecation_warning("The 'runtime_sortingfn' argument is deprecated")
+            deprecation_warning(
+                "The 'runtime_sortingfn' argument is deprecated"
+            )
         if live_dangerously is not None:
-            deprecation_warning("The 'live_dangerously' argument is deprecated")
+            deprecation_warning(
+                "The 'live_dangerously' argument is deprecated"
+            )
 
         # Parse the move schedule
         if moves is None:
@@ -237,7 +243,9 @@ class EnsembleSampler(object):
         state = State(initial_state, copy=True)
         if np.shape(state.coords) != (self.nwalkers, self.ndim):
             raise ValueError("incompatible input dimensions")
-        if (not skip_initial_state_check) and (not walkers_independent(state.coords)):
+        if (not skip_initial_state_check) and (
+            not walkers_independent(state.coords)
+        ):
             raise ValueError(
                 "Initial state has a large condition number. "
                 "Make sure that your walkers are linearly independent for the "
@@ -250,7 +258,8 @@ class EnsembleSampler(object):
         # it's current state.
         if rstate0 is not None:
             deprecation_warning(
-                "The 'rstate0' argument is deprecated, use a 'State' " "instead"
+                "The 'rstate0' argument is deprecated, use a 'State' "
+                "instead"
             )
             state.random_state = rstate0
         self.random_state = state.random_state
@@ -259,7 +268,8 @@ class EnsembleSampler(object):
         # now.
         if log_prob0 is not None:
             deprecation_warning(
-                "The 'log_prob0' argument is deprecated, use a 'State' " "instead"
+                "The 'log_prob0' argument is deprecated, use a 'State' "
+                "instead"
             )
             state.log_prob = log_prob0
         if blobs0 is not None:
@@ -312,7 +322,9 @@ class EnsembleSampler(object):
             map_fn = self.pool.map
         else:
             map_fn = map
-        model = Model(self.log_prob_fn, self.compute_log_prob, map_fn, self._random)
+        model = Model(
+            self.log_prob_fn, self.compute_log_prob, map_fn, self._random
+        )
 
         # Inject the progress bar
         total = iterations * yield_step
@@ -407,7 +419,9 @@ class EnsembleSampler(object):
                 map_func = self.pool.map
             else:
                 map_func = map
-            results = list(map_func(self.log_prob_fn, (p[i] for i in range(len(p)))))
+            results = list(
+                map_func(self.log_prob_fn, (p[i] for i in range(len(p))))
+            )
 
         try:
             log_prob = np.array([float(l[0]) for l in results])

--- a/src/emcee/ensemble.py
+++ b/src/emcee/ensemble.py
@@ -11,7 +11,7 @@ from .pbar import get_progress_bar
 from .state import State
 from .utils import deprecated, deprecation_warning
 
-__all__ = ["EnsembleSampler"]
+__all__ = ["EnsembleSampler", "walkers_independent"]
 
 try:
     from collections.abc import Iterable
@@ -78,19 +78,13 @@ class EnsembleSampler(object):
     ):
         # Warn about deprecated arguments
         if a is not None:
-            deprecation_warning(
-                "The 'a' argument is deprecated, use 'moves' instead"
-            )
+            deprecation_warning("The 'a' argument is deprecated, use 'moves' instead")
         if threads is not None:
             deprecation_warning("The 'threads' argument is deprecated")
         if runtime_sortingfn is not None:
-            deprecation_warning(
-                "The 'runtime_sortingfn' argument is deprecated"
-            )
+            deprecation_warning("The 'runtime_sortingfn' argument is deprecated")
         if live_dangerously is not None:
-            deprecation_warning(
-                "The 'live_dangerously' argument is deprecated"
-            )
+            deprecation_warning("The 'live_dangerously' argument is deprecated")
 
         # Parse the move schedule
         if moves is None:
@@ -243,14 +237,11 @@ class EnsembleSampler(object):
         state = State(initial_state, copy=True)
         if np.shape(state.coords) != (self.nwalkers, self.ndim):
             raise ValueError("incompatible input dimensions")
-        if (not skip_initial_state_check) and _scaled_cond(
-            np.atleast_2d(np.cov(state.coords, rowvar=False))
-        ) > 1e8:
-            warnings.warn(
+        if (not skip_initial_state_check) and (not walkers_independent(state.coords)):
+            raise ValueError(
                 "Initial state has a large condition number. "
                 "Make sure that your walkers are linearly independent for the "
-                "best performance",
-                category=RuntimeWarning,
+                "best performance"
             )
 
         # Try to set the initial value of the random number generator. This
@@ -259,8 +250,7 @@ class EnsembleSampler(object):
         # it's current state.
         if rstate0 is not None:
             deprecation_warning(
-                "The 'rstate0' argument is deprecated, use a 'State' "
-                "instead"
+                "The 'rstate0' argument is deprecated, use a 'State' " "instead"
             )
             state.random_state = rstate0
         self.random_state = state.random_state
@@ -269,8 +259,7 @@ class EnsembleSampler(object):
         # now.
         if log_prob0 is not None:
             deprecation_warning(
-                "The 'log_prob0' argument is deprecated, use a 'State' "
-                "instead"
+                "The 'log_prob0' argument is deprecated, use a 'State' " "instead"
             )
             state.log_prob = log_prob0
         if blobs0 is not None:
@@ -323,9 +312,7 @@ class EnsembleSampler(object):
             map_fn = self.pool.map
         else:
             map_fn = map
-        model = Model(
-            self.log_prob_fn, self.compute_log_prob, map_fn, self._random
-        )
+        model = Model(self.log_prob_fn, self.compute_log_prob, map_fn, self._random)
 
         # Inject the progress bar
         total = iterations * yield_step
@@ -420,9 +407,7 @@ class EnsembleSampler(object):
                 map_func = self.pool.map
             else:
                 map_func = map
-            results = list(
-                map_func(self.log_prob_fn, (p[i] for i in range(len(p))))
-            )
+            results = list(map_func(self.log_prob_fn, (p[i] for i in range(len(p)))))
 
         try:
             log_prob = np.array([float(l[0]) for l in results])
@@ -552,7 +537,33 @@ class _FunctionWrapper(object):
             raise
 
 
+def walkers_independent(coords):
+    if not np.all(np.isfinite(coords)):
+        return False
+    C = coords - np.mean(coords, axis=0)[None, :]
+    C_colmax = np.amax(np.abs(C), axis=0)
+    if np.any(C_colmax == 0):
+        return False
+    C /= C_colmax
+    C_colsum = np.sqrt(np.sum(C ** 2, axis=0))
+    C /= C_colsum
+    return np.linalg.cond(C.astype(float)) <= 1e8
+
+
+def walkers_independent_cov(coords):
+    C = np.cov(coords, rowvar=False)
+    if np.any(np.isnan(C)):
+        return False
+    return _scaled_cond(np.atleast_2d(C)) <= 1e8
+
+
 def _scaled_cond(a):
-    b = a / np.sqrt((a ** 2).sum(axis=0))[None, :]
-    c = b / np.sqrt((b ** 2).sum(axis=1))[:, None]
+    asum = np.sqrt((a ** 2).sum(axis=0))[None, :]
+    if np.any(asum == 0):
+        return np.inf
+    b = a / asum
+    bsum = np.sqrt((b ** 2).sum(axis=1))[:, None]
+    if np.any(bsum == 0):
+        return np.inf
+    c = b / bsum
     return np.linalg.cond(c.astype(float))

--- a/src/emcee/tests/unit/test_sampler.py
+++ b/src/emcee/tests/unit/test_sampler.py
@@ -139,7 +139,12 @@ def run_sampler(
     coords = np.random.randn(nwalkers, ndim)
     sampler = EnsembleSampler(nwalkers, ndim, normal_log_prob, backend=backend)
     sampler.run_mcmc(
-        coords, nsteps, thin=thin, thin_by=thin_by, progress=progress, store=store
+        coords,
+        nsteps,
+        thin=thin,
+        thin_by=thin_by,
+        progress=progress,
+        store=store,
     )
     return sampler
 
@@ -165,7 +170,9 @@ def test_thin(backend):
             assert np.allclose(a, c), "inconsistent {0}".format(k)
 
 
-@pytest.mark.parametrize("backend,progress", product(all_backends, [True, False]))
+@pytest.mark.parametrize(
+    "backend,progress", product(all_backends, [True, False])
+)
 def test_thin_by(backend, progress):
     with backend() as be:
         with pytest.raises(ValueError):
@@ -175,7 +182,9 @@ def test_thin_by(backend, progress):
         nsteps = 25
         thinby = 3
         sampler1 = run_sampler(None, nsteps=nsteps * thinby, progress=progress)
-        sampler2 = run_sampler(be, thin_by=thinby, progress=progress, nsteps=nsteps)
+        sampler2 = run_sampler(
+            be, thin_by=thinby, progress=progress, nsteps=nsteps
+        )
         for k in ["get_chain", "get_log_prob"]:
             a = getattr(sampler1, k)()[thinby - 1 :: thinby]
             b = getattr(sampler2, k)()

--- a/src/emcee/tests/unit/test_sampler.py
+++ b/src/emcee/tests/unit/test_sampler.py
@@ -307,7 +307,12 @@ def test_walkers_independent_scaled():
 
 @pytest.mark.parametrize(
     "nwalkers, ndim, offset",
-    [(10, 2, 1e5), (20, 5, 1e10), (30, 10, 1e14), (40, 15, 1e17)],
+    [
+        (10, 2, 1e5),
+        (20, 5, 1e10),
+        (30, 10, 1e14),
+        (40, 15, 0.1 / np.finfo(np.longdouble).eps),
+    ],
 )
 def test_walkers_independent_randn_offset_longdouble(nwalkers, ndim, offset):
     assert walkers_independent(

--- a/src/emcee/tests/unit/test_sampler.py
+++ b/src/emcee/tests/unit/test_sampler.py
@@ -6,7 +6,7 @@ from itertools import product
 import numpy as np
 import pytest
 
-from emcee import EnsembleSampler, backends, moves
+from emcee import EnsembleSampler, backends, moves, walkers_independent
 
 __all__ = ["test_shapes", "test_errors", "test_thin", "test_vectorize"]
 
@@ -116,17 +116,12 @@ def test_errors(backend, nwalkers=32, ndim=3, nsteps=5, seed=1234):
         # Ensure that a warning is logged if the inital coords don't allow
         # the chain to explore all of parameter space, and that one is not
         # if we explicitly disable it, or the initial coords can.
-        with pytest.warns(RuntimeWarning) as recorded_warnings:
+        with pytest.raises(ValueError):
             sampler.run_mcmc(np.ones((nwalkers, ndim)), nsteps)
-            assert len(recorded_warnings) == 1
-        with pytest.warns(None) as recorded_warnings:
-            sampler.run_mcmc(
-                np.ones((nwalkers, ndim)),
-                nsteps,
-                skip_initial_state_check=True,
-            )
-            sampler.run_mcmc(np.random.randn(nwalkers, ndim), nsteps)
-            assert len(recorded_warnings) == 0
+        sampler.run_mcmc(
+            np.ones((nwalkers, ndim)), nsteps, skip_initial_state_check=True
+        )
+        sampler.run_mcmc(np.random.randn(nwalkers, ndim), nsteps)
 
 
 def run_sampler(
@@ -144,12 +139,7 @@ def run_sampler(
     coords = np.random.randn(nwalkers, ndim)
     sampler = EnsembleSampler(nwalkers, ndim, normal_log_prob, backend=backend)
     sampler.run_mcmc(
-        coords,
-        nsteps,
-        thin=thin,
-        thin_by=thin_by,
-        progress=progress,
-        store=store,
+        coords, nsteps, thin=thin, thin_by=thin_by, progress=progress, store=store
     )
     return sampler
 
@@ -175,9 +165,7 @@ def test_thin(backend):
             assert np.allclose(a, c), "inconsistent {0}".format(k)
 
 
-@pytest.mark.parametrize(
-    "backend,progress", product(all_backends, [True, False])
-)
+@pytest.mark.parametrize("backend,progress", product(all_backends, [True, False]))
 def test_thin_by(backend, progress):
     with backend() as be:
         with pytest.raises(ValueError):
@@ -187,9 +175,7 @@ def test_thin_by(backend, progress):
         nsteps = 25
         thinby = 3
         sampler1 = run_sampler(None, nsteps=nsteps * thinby, progress=progress)
-        sampler2 = run_sampler(
-            be, thin_by=thinby, progress=progress, nsteps=nsteps
-        )
+        sampler2 = run_sampler(be, thin_by=thinby, progress=progress, nsteps=nsteps)
         for k in ["get_chain", "get_log_prob"]:
             a = getattr(sampler1, k)()[thinby - 1 :: thinby]
             b = getattr(sampler2, k)()
@@ -237,3 +223,85 @@ def test_pickle(backend):
             a = getattr(sampler1, k)()
             b = getattr(sampler2, k)()
             assert np.allclose(a, b), "inconsistent {0}".format(k)
+
+
+@pytest.mark.parametrize("nwalkers, ndim", [(10, 2), (20, 5)])
+def test_walkers_dependent_ones(nwalkers, ndim):
+    assert not walkers_independent(np.ones((nwalkers, ndim)))
+
+
+@pytest.mark.parametrize("nwalkers, ndim", [(10, 11), (2, 3)])
+def test_walkers_dependent_toofew(nwalkers, ndim):
+    assert not walkers_independent(np.random.randn(nwalkers, ndim))
+
+
+@pytest.mark.parametrize("nwalkers, ndim", [(10, 2), (20, 5)])
+def test_walkers_independent_randn(nwalkers, ndim):
+    assert walkers_independent(np.random.randn(nwalkers, ndim))
+
+
+@pytest.mark.parametrize(
+    "nwalkers, ndim, offset", [(10, 2, 1e5), (20, 5, 1e10), (30, 10, 1e14)]
+)
+def test_walkers_independent_randn_offset(nwalkers, ndim, offset):
+    assert walkers_independent(
+        np.random.randn(nwalkers, ndim) + np.ones((nwalkers, ndim)) * offset
+    )
+
+
+def test_walkers_dependent_big_offset():
+    nwalkers, ndim = 30, 10
+    offset = 10 / np.finfo(float).eps
+    assert not walkers_independent(
+        np.random.randn(nwalkers, ndim) + np.ones((nwalkers, ndim)) * offset
+    )
+
+
+def test_walkers_dependent_subtle():
+    nwalkers, ndim = 30, 10
+    w = np.random.randn(nwalkers, ndim)
+    assert walkers_independent(w)
+    # random unit vector
+    p = np.random.randn(ndim)
+    p /= np.sqrt(np.dot(p, p))
+    # project away the direction of p
+    w -= np.sum(p[None, :] * w, axis=1)[:, None] * p[None, :]
+    assert not walkers_independent(w)
+    # shift away from the origin
+    w += p[None, :]
+    assert not walkers_independent(w)
+
+
+def test_walkers_almost_dependent():
+    nwalkers, ndim = 30, 10
+    squash = 1e-8
+    w = np.random.randn(nwalkers, ndim)
+    assert walkers_independent(w)
+    # random unit vector
+    p = np.random.randn(ndim)
+    p /= np.sqrt(np.dot(p, p))
+    # project away the direction of p
+    proj = np.sum(p[None, :] * w, axis=1)[:, None] * p[None, :]
+    w -= proj
+    w += squash * proj
+    assert not walkers_independent(w)
+
+
+def test_walkers_independent_scaled():
+    # Some of these scales will overflow if squared, hee hee
+    scales = np.array([1, 1e10, 1e100, 1e200, 1e-10, 1e-100, 1e-200])
+    ndim = len(scales)
+    nwalkers = 5 * ndim
+    w = np.random.randn(nwalkers, ndim) * scales[None, :]
+    assert walkers_independent(w)
+
+
+@pytest.mark.parametrize(
+    "nwalkers, ndim, offset",
+    [(10, 2, 1e5), (20, 5, 1e10), (30, 10, 1e14), (40, 15, 1e17)],
+)
+def test_walkers_independent_randn_offset_longdouble(nwalkers, ndim, offset):
+    assert walkers_independent(
+        np.random.randn(nwalkers, ndim)
+        + np.ones((nwalkers, ndim), dtype=np.longdouble) * offset
+    )

